File: services/api/src/main/kotlin/net/blueshell/api/contribution/domain/FeeCycle.kt
package net.blueshell.api.contribution.domain

import net.blueshell.api.shared.dto.bulk.BulkFeeType
import net.blueshell.api.shared.dto.bulk.BulkRowDisposition
import net.blueshell.api.shared.dto.bulk.BulkRowReason
import net.blueshell.api.shared.dto.bulk.FeeCycleGroup
import net.blueshell.api.shared.enums.MemberType
import java.time.LocalDate

/**
 * When each side of the cycle's money moves.
 *
 * The two travel together everywhere — the request, the send, the preview — because a cycle
 * promises both or neither: whichever side a member is on, they are told a date.
 */
data class FeeCycleDates(
    /** The date the transfer group is asked to have paid by. */
    val paymentDue: LocalDate,
    /** The date the direct-debit group is told the money will be taken. */
    val debit: LocalDate,
)

/**
 * One member's place in a fee cycle: which side of the direct-debit partition they are on,
 * what they owe and why, and whether they will be written to at all.
 *
 * A row that will not be written to carries its reason rather than being left out, so a
 * member's absence from the send is visible instead of silent.
 */
data class FeeCycleParticipant(
    val userId: Long,
    val name: String,
    val memberType: MemberType,
    /** Start date of the membership every decision here was judged against. */
    val memberSince: LocalDate?,
    val group: FeeCycleGroup,
    val disposition: BulkRowDisposition,
    val reason: BulkRowReason?,
    /** Null exactly when no fee applies, which is when the member is honorary. */
    val feeType: BulkFeeType?,
    /** Derived from [feeType] and the period. Never typed by anyone. */
    val amount: Double?,
    /** When this member was last asked for this period, if they have been. */
    val lastAskedOn: LocalDate?,
) {
    /**
     * Whether the cycle will write to this member.
     *
     * `EXCLUDED` is not overridable and there is nothing else a fee cycle skips: every
     * member in the plan has not paid, which is what put them there.
     */
    val willSend: Boolean get() = disposition == BulkRowDisposition.INCLUDED
}

/**
 * The whole cycle for one period: every member of it who has not paid, partitioned.
 *
 * Preview and send are the same plan read twice, so they cannot disagree about who is
 * included or what they owe.
 */
data class FeeCyclePlan(
    val contributionPeriodId: Long,
    val participants: List<FeeCycleParticipant>,
) {
    fun group(group: FeeCycleGroup): List<FeeCycleParticipant> = participants.filter { it.group == group }

    fun byUserId(userId: Long): FeeCycleParticipant? = participants.firstOrNull { it.userId == userId }

    val recipients: List<FeeCycleParticipant> get() = participants.filter { it.willSend }
}

File: services/frontend/src/utils/feeCycle.ts
import type {FeeCycleRowResponse} from "@/services/api"
import {formatBulkDate} from "@/utils/bulkDisposition"
import {BulkFeeType, FeeCycleGroup, type BulkRow} from "@/utils/bulkRow"

/**
 * The fee cycle's rows, as the dialog reads them.
 *
 * Unlike the paid/unpaid dialogs, these are not computed in the browser: the cycle is over
 * every unpaid member of a period rather than over a selection the page already holds, and
 * the api decides it once so the preview and the send cannot disagree. This is the mapping
 * from that answer onto the row shape the shared scaffold renders.
 *
 * A straight mapping rather than the join `bulkRowsFromPreview` does for the membership
 * actions, because the row already carries the member details. `memberSince` in particular
 * is the start of the membership the api *judged* — the active one where the member has one
 * — and a browser deriving it would be a second implementation of that rule, free to
 * disagree with the one that decided the partition.
 */

/** One row, plus the side of the partition, which the scaffold renders as its own column. */
export function toBulkRow(row: FeeCycleRowResponse): BulkRow {
  return {
    userId: row.userId,
    name: row.name,
    disposition: row.disposition,
    reason: row.reason ?? undefined,
    memberType: row.memberType,
    memberSince: row.memberSince ?? undefined,
    amount: row.amount ?? null,
    recommendedFeeType: row.feeType ?? undefined,
    lastSentOn: row.lastAskedOn ?? undefined,
    group: row.group,
  }
}

export function toBulkRows(rows: FeeCycleRowResponse[]): BulkRow[] {
  return rows.map(toBulkRow)
}

export const feeCycleGroupLabels: Record<FeeCycleGroup, string> = {
  [FeeCycleGroup.DIRECT_DEBIT]: "Direct debit",
  [FeeCycleGroup.TRANSFER]: "Transfer",
}

export function feeCycleGroupLabel(group: FeeCycleGroup | undefined): string {
  return group ? feeCycleGroupLabels[group] : "—"
}

/** Two chips rather than one, because the count per group is the thing being checked. */
export function countByGroup(rows: BulkRow[]): Record<FeeCycleGroup, number> {
  return {
    DIRECT_DEBIT: rows.filter((row) => row.group === FeeCycleGroup.DIRECT_DEBIT).length,
    TRANSFER: rows.filter((row) => row.group === FeeCycleGroup.TRANSFER).length,
  }
}

/**
 * A member never asked reads as never, not as an empty cell — the treasurer is looking for
 * the ones already asked, so the absence has to be as legible as the date.
 */
export function lastAskedLabel(iso: string | undefined): string {
  const formatted = formatBulkDate(iso)
  // formatBulkDate is the one date format every bulk dialog uses; only the word for an
  // absent one differs, because here the absence is the answer rather than missing data.
  return formatted === "—" ? "Never" : formatted
}

/**
 * Whether this member has already been asked for this period on this side of the partition.
 *
 * Sending again is allowed, as often as the treasurer needs, so an already-asked member
 * stays included rather than becoming a row to tick back in — half way through the year that
 * would be a hundred ticks. It is warned about instead, and each ask is its own record, so
 * the warning is a note about history rather than a constraint.
 */
export function askedAlready(row: BulkRow): boolean {
  return !!row.lastSentOn
}

/** How many of the members about to be written to have already been asked. */
export function countAskedAlready(rows: BulkRow[]): number {
  return rows.filter((row) => row.disposition === "INCLUDED" && askedAlready(row)).length
}

/**
 * Only the fee types the treasurer actually changed.
 *
 * Sending every row's type would make the request state a choice where none was made, and
 * the api refuses a type naming somebody it does not write to — so a row that dropped out
 * of the cycle since the preview would refuse a send the treasurer did not ask to change.
 */
export function changedFeeTypes(
  rows: BulkRow[],
  selections: Record<number, BulkFeeType>,
): Record<string, BulkFeeType> {
  const changed: Record<string, BulkFeeType> = {}
  for (const row of rows) {
    const chosen = selections[row.userId]
    if (chosen && chosen !== row.recommendedFeeType) changed[String(row.userId)] = chosen
  }
  return changed
}
